File: cli/node/src/application/onboarding/agent-mode.ts
import { bootstrapAgent, sendEmailCode, verifyEmailCode } from '../../backend/auth.js';
import type { JsonObject } from '../../backend/json.js';
import { formatJsonEnvelope } from '../../output/format.js';
import { exit } from '../../runtime/exit.js';
import { isAgentMode } from '../../runtime/state.js';
/** Agent Mode 初始化与验证码认领流程。 */

import readline from 'node:readline';
import { type MemGoConfig, saveConfig } from '../../config/store.js';
import { colors, printError, printSuccess } from '../../output/branding.js';

const { brand, dim } = colors;

export interface BootstrapEnvelope {
  api_key: string;
  default_user_id: string;
  claim_command?: string;
  memgo_notice?: string;
}

/**
 * 检查代理账号响应中密钥与默认用户 ID 是否有效。
 * @param v - 待校验的接口响应。
 * @returns 响应是否包含有效密钥与默认用户 ID。
 */
function isValidEnvelope(v: unknown): v is BootstrapEnvelope {
  return (
    !!v &&
    typeof v === 'object' &&
    typeof (v as BootstrapEnvelope).api_key === 'string' &&
    (v as BootstrapEnvelope).api_key.length > 0 &&
    typeof (v as BootstrapEnvelope).default_user_id === 'string' &&
    (v as BootstrapEnvelope).default_user_id.length > 0
  );
}

/**
 * 调用初始化接口，保存返回的配置。source 为渠道，agentCaller 是调用方显式声明的代理身份。
 * @param input - 账号操作前的配置。
 * @param root0 - 账号操作选项。
 * @param root0.source - 账号创建来源。
 * @param root0.agentCaller - 代理调用方名称。
 * @returns 已保存代理账号信息的新配置。
 */
export async function bootstrapViaBackend(
  input: MemGoConfig,
  { source, agentCaller }: { source?: string | null; agentCaller?: string | null }
): Promise<MemGoConfig> {
  const config: MemGoConfig = {
    ...input,
    platform: { ...input.platform },
    defaults: { ...input.defaults },
  };
  const baseUrl = (config.platform.baseUrl || 'https://api.memgo.ai').replace(/\/+$/, '');
  const body: JsonObject = {};
  if (source) body.source = source;
  if (agentCaller) body.agent_caller = agentCaller;

  let resp: Response;
  try {
    resp = await bootstrapAgent(baseUrl, body);
  } catch (err) {
    printError(
      `Network error contacting MemGo: ${err instanceof Error ? err.message : String(err)}`
    );
    exit(1);
  }

  if (resp.status === 429) {
    printError('Rate-limited. Try again in a few minutes.');
    exit(1);
  }
  if (resp.status === 503) {
    printError('Agent Mode is temporarily disabled. Try again later.');
    exit(1);
  }
  if (!resp.ok) {
    let detail: string = resp.statusText;
    try {
      const errBody = (await resp.json()) as {
        error?: string;
        detail?: string;
      };
      detail = errBody.error ?? errBody.detail ?? resp.statusText;
    } catch {
      /* 非 JSON 错误响应保留 HTTP 状态说明。 */
    }
    // 将平台返回的通用权限拒绝映射为现有注册限流提示。

    if (resp.status === 403 && /permission/i.test(detail)) {
      printError(
        'Daily Agent Mode signup limit reached for this network (5/day). Try again from a different IP or after midnight UTC.'
      );
      exit(1);
    }
    printError(`Bootstrap failed: ${detail}`);
    exit(1);
  }

  const envelope = (await resp.json()) as BootstrapEnvelope;
  if (!isValidEnvelope(envelope)) {
    // 校验初始化响应必需字段，避免将空密钥写入配置。

    printError('Bootstrap response missing required fields — please update the CLI.');
    exit(1);
  }

  config.platform.apiKey = envelope.api_key;
  config.platform.baseUrl = baseUrl;
  config.platform.agentMode = true;
  config.platform.createdVia = 'agent_mode';
  config.platform.agentCaller = agentCaller ?? '';
  config.platform.claimedAt = '';
  config.platform.defaultUserId = envelope.default_user_id;
  // 以初始化返回的用户标识作为记忆操作默认范围。
  config.defaults.userId = envelope.default_user_id;
  saveConfig(config);
  if (isAgentMode()) {
    formatJsonEnvelope({
      command: 'init',
      data: {
        api_key_saved: true,
        agent_mode: config.platform.agentMode,
        user_id: config.defaults.userId,
      },
    });
    return config;
  }

  printSuccess(`Agent Mode active. Default user_id: ${envelope.default_user_id}`);
  if (envelope.memgo_notice) {
    console.log(`\n\x1b[33m🔔 ${envelope.memgo_notice}\x1b[0m\n`);
  } else {
    // 旧服务没有统一提示字段时保留现有认领说明。
    const claimCmd = envelope.claim_command ?? 'memgo init --email <your-email>';
    console.log(`  ${dim(`To claim this account later: ${claimCmd}`)}`);
  }

  if (!agentCaller) {
    console.log(
      `  ${dim('If you are an AI agent, identify yourself: `memgo identify <your-name>` (e.g. claude-code, cursor).')}`
    );
  }
  return config;
}

/**
 * 通过邮箱验证码原地认领账号；携带 agent_mode_api_key 验证，不创建替代账号。
 * @param input - 账号操作前的配置。
 * @param root0 - 账号操作选项。
 * @param root0.email - 用于登录或认领的邮箱。
 * @param root0.code - 邮箱验证码。
 * @returns 已认领账号的新配置。
 */
export async function claimViaOtp(
  input: MemGoConfig,
  { email, code }: { email: string; code?: string }
): Promise<MemGoConfig> {
  const config: MemGoConfig = {
    ...input,
    platform: { ...input.platform },
    defaults: { ...input.defaults },
  };
  const baseUrl = (config.platform.baseUrl || 'https://api.memgo.ai').replace(/\/+$/, '');
  if (!config.platform.apiKey || !config.platform.agentMode) {
    printError('This command requires an active Agent Mode config. Run `memgo init` first.');
    exit(1);
  }

  const rawKey = config.platform.apiKey;

  // 未提供验证码时先请求发送。
  if (!code) {
    const sendResp = await sendEmailCode(baseUrl, email);
    if (sendResp.status === 429) {
      printError('Too many attempts. Try again in a few minutes.');
      exit(1);
    }
    if (!sendResp.ok) {
      let detail: string = sendResp.statusText;
      try {
        const errBody = (await sendResp.json()) as { error?: string };
        if (errBody.error) detail = errBody.error;
      } catch {
        /* 非 JSON 错误响应保留 HTTP 状态说明。 */
      }
      printError(`Failed to send code: ${detail}`);
      exit(1);
    }

    printSuccess(`Verification code sent to ${email}. Check your inbox.`);

    if (!process.stdin.isTTY) {
      printError(
        'No --code provided and terminal is non-interactive.',
        `Re-run: memgo init --email ${email} --code <code>`
      );
      exit(1);
    }

    console.log();
    code = await promptLine(`  ${brand('Verification Code')}`);
    if (!code) {
      printError('Code is required.');
      exit(1);
    }
  }

  // 验证码验证与账号认领由服务端原子执行。
  const verifyResp = await verifyEmailCode(baseUrl, email, code.trim(), rawKey);

  if (!verifyResp.ok) {
    let detail: string = verifyResp.statusText;
    let errCode = '';
    try {
      const errBody = (await verifyResp.json()) as {
        error?: string;
        code?: string;
      };
      if (errBody.error) detail = errBody.error;
      if (errBody.code) errCode = errBody.code;
    } catch {
      /* 非 JSON 错误响应保留 HTTP 状态说明。 */
    }
    printError(`Claim failed: ${detail}`);
    if (errCode === 'email_already_claimed') {
      console.log(
        `  ${dim('Tip: this email already has a MemGo account. Sign in at app.memgo.ai with your existing credentials.')}`
      );
    }
    exit(1);
  }

  const claimBody = (await verifyResp.json()) as {
    claimed?: boolean;
    claimed_at?: string;
  };
  if (!claimBody.claimed) {
    printError('Unexpected verify response: claimed must be true');
    exit(1);
  }

  config.platform.agentMode = false;
  config.platform.claimedAt = claimBody.claimed_at ?? new Date().toISOString();
  config.platform.userEmail = email;
  config.platform.createdVia = 'email';
  saveConfig(config);
  if (isAgentMode()) {
    formatJsonEnvelope({
      command: 'init',
      data: {
        api_key_saved: true,
        agent_mode: config.platform.agentMode,
        user_id: config.defaults.userId,
      },
    });
    return config;
  }

  printSuccess(`Agent claimed to ${email}. Your API key is unchanged.`);
  return config;
}

/**
 * 读取一行终端输入并清除首尾空白。
 * @param label - 终端提示文本。
 * @returns 清除首尾空白后的输入；支持默认值的入口在空输入时使用默认值。
 */
function promptLine(label: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  return new Promise((resolve) => {
    rl.question(`${label}: `, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

File: cli/python/src/memgo_cli/application/onboarding/agent_mode.py
"""代理账号创建和邮箱验证码认领用例。"""

from __future__ import annotations

import json
import sys
from copy import deepcopy
from datetime import datetime, timezone

import httpx
import typer
from rich.console import Console
from rich.prompt import Prompt

from memgo_cli.backend.auth import bootstrap_agent, send_email_code, verify_email_code
from memgo_cli.backend.json import JsonObject, json_object, json_string
from memgo_cli.config.models import MemGoConfig
from memgo_cli.config.store import save_config
from memgo_cli.output.branding import (
    BRAND_COLOR,
    DIM_COLOR,
    print_error,
    print_success,
)

console = Console()
err_console = Console(stderr=True)


def _validate_envelope(envelope: object) -> None:
    """拒绝缺少非空密钥或默认用户 ID 的初始化响应。"""
    if not isinstance(envelope, dict):
        print_error(err_console, "Bootstrap response was not a JSON object.", hint=None)
        raise typer.Exit(1)
    for field in ("api_key", "default_user_id"):
        value = envelope.get(field)
        if not isinstance(value, str) or not value:
            print_error(
                err_console,
                f"Bootstrap response missing required field {field!r} — please update the CLI.",
                hint=None,
            )
            raise typer.Exit(1)


def bootstrap_via_backend(
    config: MemGoConfig,
    *,
    source: str | None,
    agent_caller: str | None,
) -> None:
    """创建代理账号并保存独立配置副本。

    source 和 agent_caller 按原协议传递；输入配置不修改，失败时退出。
    """
    config = deepcopy(config)
    base_url = (config.platform.base_url or "https://api.memgo.ai").rstrip("/")
    body: JsonObject = {}
    if source:
        body["source"] = source
    if agent_caller:
        body["agent_caller"] = agent_caller

    try:
        resp = bootstrap_agent(base_url, body)
    except httpx.HTTPError as exc:
        print_error(err_console, f"Network error contacting MemGo: {exc}", hint=None)
        raise typer.Exit(1) from exc

    if resp.status_code == 429:
        print_error(err_console, "Rate-limited. Try again in a few minutes.", hint=None)
        raise typer.Exit(1)
    if resp.status_code == 503:
        print_error(err_console, "Agent Mode is temporarily disabled. Try again later.", hint=None)
        raise typer.Exit(1)
    if resp.status_code != 200:
        detail = resp.text
        try:
            err_body = resp.json()
            detail = err_body.get("error") or err_body.get("detail") or resp.text
        except (json.JSONDecodeError, ValueError, AttributeError):
            pass
        # 后端限流装饰器抛出 PermissionDenied
        # DRF 将其转换为通用 403 权限错误
        # 此处按既有协议显示每日注册限额提示
        if resp.status_code == 403 and "permission" in str(detail).lower():
            print_error(
                err_console,
                "Daily Agent Mode signup limit reached for this network (5/day). Try again from a different IP or after midnight UTC.",
                hint=None,
            )
            raise typer.Exit(1)
        print_error(err_console, f"Bootstrap failed: {detail}", hint=None)
        raise typer.Exit(1)

    envelope = resp.json()
    _validate_envelope(envelope)
    config.platform.api_key = envelope["api_key"]
    config.platform.base_url = base_url
    config.platform.agent_mode = True
    config.platform.created_via = "agent_mode"
    config.platform.agent_caller = agent_caller or ""
    config.platform.claimed_at = ""
    config.platform.default_user_id = envelope["default_user_id"]
    # 将初始化返回的用户 ID 作为默认记忆范围
    config.defaults.user_id = envelope["default_user_id"]
    save_config(config)

    from memgo_cli.output.format import format_json_envelope
    from memgo_cli.runtime.state import capture_notice, is_agent_mode

    if is_agent_mode():
        notice = envelope.get("memgo_notice")
        capture_notice(None if notice is None else json_string(notice, "memgo_notice"))
        format_json_envelope(
            console,
            command="init",
            data={
                "api_key_saved": True,
                "agent_mode": True,
                "default_user_id": envelope["default_user_id"],
            },
            duration_ms=None,
            scope=None,
            count=None,
            status="success",
            error=None,
        )
        return

    print_success(console, f"Agent Mode active. Default user_id: {envelope['default_user_id']}")
    notice = envelope.get("memgo_notice")
    if notice:
        console.print(f"\n[yellow]🔔 {notice}[/yellow]\n")
    else:
        # 兼容尚未返回统一通知的后端版本
        claim_cmd = envelope.get("claim_command", "memgo init --email <your-email>")
        console.print(f"  [{DIM_COLOR}]To claim this account later: {claim_cmd}[/]")

    if not agent_caller:
        console.print(
            f"  [{DIM_COLOR}]If you are an AI agent, identify yourself: "
            f"`memgo identify <your-name>` (e.g. claude-code, cursor).[/]"
        )


def claim_via_otp(config: MemGoConfig, *, email: str, code: str | None) -> None:
    """通过邮箱验证码认领代理账号并保存配置副本。

    验证请求携带原代理密钥；成功后更新认领状态和邮箱，密钥不变。
    """
    config = deepcopy(config)
    base_url = (config.platform.base_url or "https://api.memgo.ai").rstrip("/")
    if not config.platform.api_key or not config.platform.agent_mode:
        print_error(
            err_console,
            "This command requires an active Agent Mode config. Run `memgo init` first.",
            hint=None,
        )
        raise typer.Exit(1)

    raw_key = config.platform.api_key

    if not code:
        send = send_email_code(base_url, email)
        if send.status_code == 429:
            print_error(err_console, "Too many attempts. Try again in a few minutes.", hint=None)
            raise typer.Exit(1)
        if send.status_code != 200:
            try:
                detail = send.json().get("error", send.text)
            except (ValueError, AttributeError):
                detail = send.text
            print_error(err_console, f"Failed to send code: {detail}", hint=None)
            raise typer.Exit(1)

        print_success(console, f"Verification code sent to {email}. Check your inbox.")

        if not sys.stdin.isatty():
            print_error(
                err_console,
                "No --code provided and terminal is non-interactive.",
                hint=f"Re-run: memgo init --email {email} --code <code>",
            )
            raise typer.Exit(1)

        console.print()
        code = Prompt.ask(f"  [{BRAND_COLOR}]Verification Code[/]")
        if not code:
            print_error(err_console, "Code is required.", hint=None)
            raise typer.Exit(1)

    # 一次请求完成验证码验证和账号认领
    verify = verify_email_code(base_url, email, code, raw_key)

    if verify.status_code != 200:
        try:
            err_body = verify.json()
            detail = err_body.get("error", verify.text)
            code_str = err_body.get("code", "")
        except (json.JSONDecodeError, ValueError, AttributeError):
            detail = verify.text
            code_str = ""
        print_error(err_console, f"Claim failed: {detail}", hint=None)
        if code_str == "email_already_claimed":
            console.print(
                f"  [{DIM_COLOR}]Tip: this email already has a MemGo account. Sign in at app.memgo.ai with your existing credentials.[/]"
            )
        raise typer.Exit(1)

    claim_body = json_object(verify.json())
    if claim_body.get("claimed") is not True:
        print_error(err_console, "Unexpected verify response: claimed must be true", hint=None)
        raise typer.Exit(1)

    config.platform.agent_mode = False
    config.platform.claimed_at = json_string(
        claim_body.get("claimed_at") or _utcnow_iso(), "claimed_at"
    )
    config.platform.user_email = email
    config.platform.created_via = "email"
    save_config(config)
    from memgo_cli.output.format import format_json_envelope
    from memgo_cli.runtime.state import is_agent_mode

    if is_agent_mode():
        format_json_envelope(
            console,
            command="init",
            data={"claimed": True, "agent_mode": False, "user_email": email},
            duration_ms=None,
            scope=None,
            count=None,
            status="success",
            error=None,
        )
        return
    print_success(console, f"Agent claimed to {email}. Your API key is unchanged.")


def _utcnow_iso() -> str:
    """返回当前 UTC 时间的 ISO 文本。"""
    return datetime.now(timezone.utc).isoformat()
